File: wallet-kit/src/main/java/bitcoin/walllet/kit/network/message/PongMessage.java
package bitcoin.walllet.kit.network.message;

import java.io.ByteArrayInputStream;
import java.io.IOException;

import bitcoin.walllet.kit.common.io.BitcoinInput;
import bitcoin.walllet.kit.common.io.BitcoinOutput;

/**
 * Build P2P message:
 * https://en.bitcoin.it/wiki/Protocol_documentation#Message_structure
 *
 * @author liaoxuefeng
 */
public class PongMessage extends Message {

    long nonce;

    public PongMessage(long nonce) {
        super("pong");
        this.nonce = nonce;
    }

    public PongMessage(byte[] payload) throws IOException {
        super("pong");
        try (BitcoinInput input = new BitcoinInput(new ByteArrayInputStream(payload))) {
            this.nonce = input.readLong();
        }
    }

    @Override
    protected byte[] getPayload() {
        return new BitcoinOutput().writeLong(this.nonce).toByteArray();
    }

    @Override
    public String toString() {
        return "PongMessage(nonce=" + this.nonce + ")";
    }

}

File: wallet-kit/src/main/kotlin/bitcoin/wallet/kit/messages/PongMessage.kt
package bitcoin.wallet.kit.messages

import bitcoin.walllet.kit.common.io.BitcoinInput
import bitcoin.walllet.kit.common.io.BitcoinOutput
import java.io.ByteArrayInputStream
import java.io.IOException

/**
 * Pong Message
 *
 *  Size        Field   Description
 *  ====        =====   ===========
 *  8 bytes     Nonce   Random value
 */
class PongMessage : Message {

    private var nonce: Long = 0

    constructor(nonce: Long) : super("pong") {
        this.nonce = nonce
    }

    @Throws(IOException::class)
    constructor(payload: ByteArray) : super("pong") {
        BitcoinInput(ByteArrayInputStream(payload)).use { input ->
            nonce = input.readLong()
        }
    }

    override fun getPayload(): ByteArray {
        return BitcoinOutput().writeLong(nonce).toByteArray()
    }

    override fun toString(): String {
        return "PongMessage(nonce=$nonce)"
    }
}
